utils/kmgenc/kmgenc.1
.TH KMGENC 1 "Mar 2017" "Version 1.0"
.SH NAME
kmgenc \- Store images as SEGA Dreamcast PVR textures in a KMG container
.SH SYNOPSIS
\fBkmgenc\fR [\fIOPTION\fR]... [\fIFILE\fR]...

.SH DESCRIPTION
.B kmgenc
is used to convert images to the SEGA Dreamcast's available
image formats. Typically this means RGB565, ARGB1555 or ARGB4444.

.SH OPTIONS
.TP
.BR \-v ", " \-\-verbose\fR
Print more information during processing.

.TP
.BR \-d ", " \-\-debug\fR
Print much more information.

.TP
.BR \-a4 ", " \-\-argb4444\fR
Use 4 bit alpha channel (Dreamcast PVR texture format ARGB4444).

.TP
.BR \-a1 ", " \-\-argb1555\fR
Use 1 bit alpha channel (Dreamcast PVR texture format ARGB1555).

.SH EXAMPLES

.EX
.B
   kmgenc -a image.png
.EE

.SH AUTHOR
This manual page was initially written by Stefan Galowicz <bogglez@protonmail.ch>,
for the KOS project.
.TP
The program has been initially written by Gil Megidish in 2002 and modified by
Megan Potter in 2003.

.SH BUGS
kmgenc is very incomplete and you may want to use vqenc instead, since it is also
capable of outputting kmg container files with mipmaps and VQ compression.


doc/CHANGELOG
Platform independent changes are marked with '***', otherwise the
applicable platform is listed. The name/initials of the person who
is responsible for the change are listed in [square brackets] for most
versions (where they are missing from early versions, assume that they
should be attributed to Megan Potter).

KallistiOS version 2.1.0 -----------------------------------------------
- *** Cleaned up generated stubs files on a make clean [Lawrence Sebald == LS]
- DC  Added a function to detect if the program is being run on a retail
      Dreamcast or a Set5.xx device [LS]
- DC  Fixed an issue with the SIP driver that would cause programs to freeze
      if a microphone was connected at program startup [LS]
- DC  Rearranged the SIP driver to get rid of the internal buffer [LS]
- *** Removed the GBA, ia32, and ps2 ports. If someone really wants them back
      and wants to maintain them, let me know. [LS]
- DC  Fixed fmath functions so they compile in C99 mode [LS]
- *** Cleaned up warnings with -std=gnu99 for C code [LS]
- *** Cleaned up warnings with -std=c99 for C code [LS]
- DC  Changed the PVR render-to-texture support to allow for two render-to-
      texture passes without an intervening render-to-screen pass [LS]
- *** Fixed an issue in fs_ext2 that would allow you to rename a directory to
      be a child of itself, thus chopping it off from the rest of the
      directory tree [LS]
- *** Added a fs_readlink() function to resolve symlinks [LS]
- *** Fixed the vqenc and kmgenc utilities for compilation on a 64-bit OS [LS]
- DC  Corrected an issue with video cable detection with GCC 4.8.x [LS]
- DC  Added support for ATA devices on the G1 bus [LS]
- DC  Fixed an infinite loop in ATA initialization if nothing except the
      GD-ROM drive is connected [LS]
- *** Added documentation for the addons tree [LS]
- DC  Made the hardware auto-initialization/shutdown functions weakly linked so
      that they can be overridden without modifying KOS [LS]
- DC  Corrected render-to-texture mode so that you don't have to wait for a
      vblank to actually render to the texture [LS]
- *** Modified fs_stat to conform (mostly) to the POSIX standard for the stat
      function [LS]
- DC  Modified the cdrom driver so that it will set itself as the active ATA
      device before trying to do any CD related reading [LS]
- DC  Fixed potential race conditions between the cdrom code and g1ata [LS]
- DC  Fixed full sector reading of non-data CDs (thanks to SWAT for pointing
      out that it needed fixing) [LS]
- DC  Added many new matrix/vector math-related macros [Josh Pearson == JP]
- *** Added libppp for dialup internet access to the addons tree [LS]
- *** Added DNS resolver functions getaddrinfo() and gethostbyname() [LS]
- *** Fixed a problem where poll() would deadlock if the timeout expired [LS]
- *** Modified getaddrinfo() so that it would try multiple times to contact
      the DNS server with a timeout between messages [LS]
- DC  Various cleanups and modifications to the cdrom code, including subcode
      reading functionality, DMA-based sector reads, and using a recursive
      mutex for the G1 lock [Donald Haase == DH]
- DC  Fixed pvr_mem_available() to report all available texture memory, not
      just what was available in the pvr_malloc() arena [LS]
- DC  Added a utility function to the PVR code for controlling the cheap shadow
      scale register [LS]
- DC  Added an example of the PVR's cheap shadow functionality [LS]
- DC  Added a tool for generation of bumpmaps that was sent to the mailing list
      all the way back in 2005 [Fredrik Ehnbom == FE]
- *** Added the fs_path_append() function to append path components [LS]
- *** Added a nanosleep() function to koslib by request [LS]
- *** Added rewinddir() support to the VFS and serveral filesystems [LS]
- *** Added in C11 threading support [LS]
- *** Added UDP Lite support to the network stack [LS]
- DC  Adjust sq_cpy() to use a const pointer for the source [JP]
- DC  Retrieve the IP address from dcload-ip when we can [LS]
- *** Added the mutex_unlock_as_thread() function to allow an IRQ handler to
      unlock a mutex as if it were a specified thread [LS]
- DC  Fixed a deadlock issue in the G1ATA code related to the use of a recursive
      mutex for locking -- Thanks to SWAT for noticing the issue [LS]
- DC  Added support for German (DE) and UK keyboards [Christian Groessler == CG]
- DC  Minor cleanups to the flashrom code [DH && LS]
- DC  Added light gun support roughly based on how it is done in libronin [LS]
- DC  Added a very simple light gun example [LS]
- DC  Added support for Spanish keyboards [Xavier Vallejo == XV]
- DC  Added rudimentary support for OCRAM in linker script [CG]
- *** Added the C11 aligned_alloc() function [LS]
- *** Added fs_fstat and modified the fstat function in the newlib handlers to
      use that or fake it for filesystems that don't support fstat [LS]
- DC  Added fstat support to fs_iso9660 and fs_vmu [LS]
- *** Added fstat support to fs_ext2, fs_ramdisk, fs_romdisk, and fs_pty [LS]
- *** Changed libc/koslib/byteorder.c functions to use the <arch/byteorder.h>
      macros, instead of being hard-coded as little-endian [LS]
- DC  Added Opus decoding example using libopusplay [LS]
- *** Removed jiffy "counter" that wasn't actually counting anything [LS]
- *** Added the thd_create_ex() function to create a thread with a given set of
      attributes set at thread creation (allowing us to expand that list of
      attributes a bit more easily than what we would have to do otherwise) [LS]
- *** Made thd_mode local to the threading code (if you need to get the mode at
      runtime, for some reason, use the new thd_get_mode() function) [LS]
- DC  Patched the GCC patch so that GCC 4.7.x compiles cleanly with a newer host
      GCC version [Luke Benstead == LB]
- *** Fixed GCC version detection in the environ script for some locales [LS]

utils/kmgenc/kmgenc.c
/* KallistiOS ##version##

   kmgenc.c
   Copyright (C)2003 Megan Potter
   Portions Copyright (C)2002 Gil Megidish

   This utility loads JPGs and PNGs and converts them into KMGs of the
   specified format. Right now only Dreamcast formats are supported.
   KMGs are basically just raw texture data that can be DMA'd straight
   to the PVR's memory. This is useful to improve loading times, among
   other things.

   The following options are available:
   - Twiddling (or no)
   - Mipmaps (or no)
   - VQ encoding (or no)

   Any combination of these attributes may be selected for the final
   output file. Note that input textures must be a power of 2 on each
   side, though they may not necessarily need to be square. VQ'd textures
   need to be square (supposedly).

   The following output formats are available as well:
   - RGB565
   - ARGB4444
   - ARGB1555

   The ARGB formats are not particularly useful unless you're using
   a source PNG with an alpha channel. If you select an RGB format with
   such a file, the alpha channel will be silently ignored (you may get
   background noise depending on your PNG creation program).


   XXX
   Note: this is pretty darned unfinished and needs to be combined with
   vqenc to be complete. For now (since vqenc outputs KMGs as well) I've
   just put in the basic twiddling operations.

*/

#include "kmgenc.h"

int use_twiddle = 1;
int use_verbose = 1;
int use_debug = 1;
int use_alpha = 0;

/* Linear/iterative twiddling algorithm from Marcus' tatest */
#define TWIDTAB(x) ( (x&1)|((x&2)<<1)|((x&4)<<2)|((x&8)<<3)|((x&16)<<4)| \
                     ((x&32)<<5)|((x&64)<<6)|((x&128)<<7)|((x&256)<<8)|((x&512)<<9) )
#define TWIDOUT(x, y) ( TWIDTAB((y)) | (TWIDTAB((x)) << 1) )
#define MIN(a, b) ( (a)<(b)? (a):(b) )

/* This twiddling code is copied from pvr_texture.c, and the original
   algorithm was written by Vincent Penne. */
static void twiddle(image_t * src, uint16 * output) {
    int w = src->w;
    int h = src->h;
    int min = MIN(w, h);
    int mask = min - 1;
    uint16 * pixels = (uint16 *)src->data;
    uint16 * vtex = output;
    int x, y, yout;

    for(y = 0; y < h; y++) {
        /* yout = ((h - 1) - y); */
        yout = y;

        for(x = 0; x < w; x++) {
            vtex[TWIDOUT(x & mask, yout & mask) +
                 (x / min + yout / min)*min * min] = pixels[y * w + x];
        }
    }
}

static void convert_to_16(image_t * img) {
    int i;
    fcolor_t fc;
    uint16 * out;

    out = malloc(img->w * img->h * 2);

    for(i = 0; i < img->w * img->h; i++) {
        get_fcolor_32(&fc, img->data + i * img->bpp);
        LIMITFC(fc);

        switch(use_alpha) {
            case 0:
                out[i] = le16(PACK565(fc.r, fc.g, fc.b));
                break;
            case 1:
                out[i] = le16(PACK4444(fc.a, fc.r, fc.g, fc.b));
                break;
            case 2:
                out[i] = le16(PACK1555(fc.a, fc.r, fc.g, fc.b));
                break;
        }
    }

    img->bpp = 2;
    free(img->data);
    img->data = (uint8 *)out;
}

static int save(const char *filename, image_t *img) {
    FILE    *fp;
    kmg_header_t    hdr;
    uint16      * tmp = NULL;
    int     fmt, cnt;

    fp = fopen(filename, "wb");

    if(fp == NULL) {
        fprintf(stderr, "FATAL: cannot create %s\n", filename);
        return -errno;
    }

    memset(&hdr, 0, sizeof(hdr));
    hdr.magic = le32(KMG_MAGIC);
    hdr.version = le32(KMG_VERSION);
    hdr.platform = le32(KMG_PLAT_DC);

    switch(use_alpha) {
        case 0:
            fmt = KMG_DCFMT_RGB565;
            break;
        case 1:
            fmt = KMG_DCFMT_ARGB4444;
            break;
        case 2:
            fmt = KMG_DCFMT_ARGB1555;
            break;
        default:
            exit(-1);
    }

    if(use_twiddle)
        fmt |= KMG_DCFMT_TWIDDLED;

    hdr.format = le32(fmt);
    hdr.width = le32(img->w);
    hdr.height = le32(img->h);
    cnt = img->w * img->h * 2;
    hdr.byte_count = le32(cnt);

    if(fwrite(&hdr, sizeof(hdr), 1, fp) != 1) {
        fprintf(stderr, "FATAL: can't write KMG header to %s\n", filename);
        goto loser;
    }

    /* Twiddle the image into a temp buffer */
    tmp = malloc(cnt);
    twiddle(img, tmp);

    /* Write it out */
    if(fwrite(tmp, cnt, 1, fp) != 1) {
        fprintf(stderr, "FATAL: can't write KMG data to %s\n", filename);
        goto loser;
    }

    free(tmp);
    fclose(fp);
    return 0;

loser:

    if(tmp)
        free(tmp);

    fclose(fp);
    unlink(filename);
    return -1;
}


static void banner(const char *progname) {
    printf("Usage: %s [options] image1 [image2..]\n", progname);
    printf("\n");
    printf("Options:\n");
    /* printf("\t-t, --twiddle\tcreate twiddled textures\n"); */
    /* printf("\t-m, --mipmap\tcreate mipmapped textures (EXPERIMENTAL)\n"); */
    printf("\t-v, --verbose\tverbose\n");
    printf("\t-d, --debug\tshow debug information\n");
    /* printf("\t-q, --highq\thigher quality (much slower)\n"); */
    printf("\t-a4, --argb4444\tuse alpha channel (and output ARGB4444)\n");
    printf("\t-a1, --argb1555\tuse alpha channel (and output ARGB1555)\n");
}

static int valid_size(int x) {
    if(x < 2 || x > 1024)
        return 0;

    while(x) {
        int bit = (x & 1);
        x >>= 1;

        if(x && bit) {
            /* more than one bit in image */
            return 0;
        }
    }

    return 1;
}

static const char *figure_outfilename(const char *f, const char *newext) {
    char *newname;
    char *ext;

    ext = strrchr(f, '.');

    if(ext == NULL) {
        newname = (char *)malloc(strlen(f) + strlen(newext) + 2);

        if(newname) {
            sprintf(newname, "%s.%s", f, newext);
        }
    }
    else {
        int len;

        len = (ext - f) + strlen(newext) + 2;
        newname = (char *)malloc(len);

        if(newname) {
            strcpy(newname, f);
            ext = strrchr(newname, '.') + 1;
            strcpy(ext, newext);
        }
    }

    return newname;
}

static void destroy_image(image_t *image) {
    if(image->data) {
        free(image->data);
        image->data = NULL;
    }
}

static int encode(const char *infile) {
    int     ok;
    image_t     image;
    const char  *outfile;

    if(use_verbose) {
        printf("encoding %s.. ", infile);
    }

    if(get_image(infile, &image) < 0) {
        fprintf(stderr, "failed reading %s\n", infile);
        return -EINVAL;
    }

    if(valid_size(image.w) == 0 || valid_size(image.h) == 0) {
        fprintf(stderr, "image dimensions for %s are not valid, see manual\n", infile);
        destroy_image(&image);
        return -EINVAL;
    }

    outfile = figure_outfilename(infile, "kmg");

    if(outfile == NULL) {
        fprintf(stderr, "memory allocation failed for %s\n", infile);
        return -ENOMEM;
    }

    /* Convert the input image to a 16-bit image according to parameters */
    convert_to_16(&image);

    /* Save it */
    ok = save(outfile, &image);

    destroy_image(&image);

    printf("\n");
    return ok;
}

static int process_long_options(char *arg) {
    /* if (! strcmp(arg, "mipmap"))
        use_mipmap = 1;
    else if (! strcmp(arg, "twiddle"))
        use_twiddle = 1;
    else */ if(! strcmp(arg, "debug"))
        use_debug = 1;
    else if(! strcmp(arg, "verbose"))
        use_verbose = 1;
    /* else if (! strcmp(arg, "highq"))
        use_hq = 1; */
    else if(! strcmp(arg, "alpha"))
        use_alpha = 1;
    else
        return -EINVAL;

    return 0;
}

static int process_option(char *arg) {
    /* assuming starts with '-' */
    arg++;

    switch(*arg) {

            /* case 'm':
                use_mipmap = 1;
                return 0; */

            /* case 't':
                use_twiddle = 1;
                return 0; */

        case 'v':
            use_verbose = 1;
            return 0;

        case 'd':
            use_debug = 1;
            return 0;

            /* case 'q':
                use_hq = 1;
                return 0; */

        case 'a':
            arg++;

            if(*arg == '4')
                use_alpha = 1;
            else if(*arg == '1')
                use_alpha = 2;
            else
                return -EINVAL;

            return 0;

        case '-':
            return process_long_options(arg + 1);
    }

    return -EINVAL;
}

static int process(int argc, char *argv[]) {
    int arg;

    arg = 1;

    while(arg < argc) {
        if(argv[arg][0] == '-') {
            if(process_option(argv[arg]) < 0) {
                fprintf(stderr, "invalid option %s\n", argv[arg]);
                return -EINVAL;
            }

            arg++;
            continue;
        }

        /* found a filename */
        break;
    }

    if(arg >= argc) {
        fprintf(stderr, "no files to encode\n");
        return -EINVAL;
    }

    while(arg < argc) {
        /* ordinary image */
        encode(argv[arg]);
        arg++;
    }

    return 0;
}

int main(int argc, char *argv[]) {
    setbuf(stdout, 0);

    if(argc < 2) {
        banner(argv[0]);
        return 0;
    }

    return process(argc, argv);
}
